File: tldjfj123/Week06/문자열나누기/Solution.java
package 문자열나누기;

import java.util.*;

class Solution {
    public int solution(String s) {
        //Convert String to Deque with ArrayDeque for get deque's index
        Deque<Character> arr = new ArrayDeque<>();
        for (char c : s.toCharArray()) {
            arr.offer(c);
        }

        int result = 0;

        //std = first idx value of ArrayDeque
        //Compare the first index value with the remaining values and count at sameCnt & diffCnt
        while (!arr.isEmpty()) {
            char std = arr.peekFirst();
            int sameCnt = 0;
            int diffCnt = 0;

            for (char c : arr) {
                if (std == c) {
                    sameCnt++;
                } else {
                    diffCnt++;
                }
                //If sameCnt == diffCnt -> Remove the corresponding values from the deque.
                if (sameCnt == diffCnt) {
                    result++;
                    for (int i = 0; i < sameCnt + diffCnt; i++) {
                        arr.pollFirst();
                    }
                    break;
                }
            }
            // Handle exceptions -> case "abracadabra"
            if (sameCnt != diffCnt) {
                result++;
                break;
            }
        }

        return result;
    }

}

File: tldjfj123/Week06/문자열나누기/Solution.py
from collections import deque

def solution(s):
    arr = deque(list(s))
    result = 0

    while len(arr) > 0:
        std = arr[0]
        same_cnt = 0
        diff_cnt = 0
        
        for i in range(len(arr)):
            if std == arr[i]:
                same_cnt += 1
            else:
                diff_cnt += 1
            
            if same_cnt == diff_cnt:
                result += 1
                for _ in range(same_cnt + diff_cnt):
                    arr.popleft()
                break
        
        if same_cnt != diff_cnt:
            result += 1
            break

    return result